### PR TITLE
Update new reset password flow

### PR DESCRIPTION
**Pull request by hoan-bitholla**

### CODE REVIEW

File: tools/security.js [hunk 1]
```diff
@@ -102,13 +99,19 @@ const resetUserPassword = (resetPasswordCode, newPassword) => {
...
+				})
+				.catch(() => {
+					throw new Error();
+				});
+		})
+		.then((user_id) => {
+			return dbQuery.findOne('user', { where: { id: user_id } });
 		})
-		.then((code) => dbQuery.findOne('user', { where: { id: code.user_id } }))
 		.then((user) => user.update({ password: newPassword }, { fields: ['password'] }));
```
Comment (brandonkimmmm) [inline tools/security.js:115]: We can use bluebird all here to run `client.delAsync` and `dbQuery.findOne(...` at the same time. If `client.delAsync` fails, the function will throw an error.

```
    return all([
        dbQuery.findOne('user', { where: { id: user_id } }),
        client.delAsync(`ResetPasswordCode:${resetPasswordCode}`)
    ])
})
.then(([ user ]) => {
    ...
})
```

File: tools/security.js [hunk 2]
```diff
@@ -149,21 +150,26 @@ const getResetPasswordCode = (code) => {
...
-			});
-		})
-		.then((code) => code.code);
+	const resetPasswordCodeKey = `resetPasswordCode:${userId}`;
+
+	//Generate new random code
+	const newCode = crypto.randomBytes(20).toString('hex');
+
+	//Code is expire in 24 hours
+	const setCode = client.setex(resetPasswordCodeKey, 60 * 60 * 24, newCode);
```
Comment (abeikverdi) [inline tools/security.js]: This should be shorter. set it to 5 mins.

File: tools/security.js [hunk 3]
```diff
@@ -152,8 +152,8 @@ const createResetPasswordCode = (userId) => {
 	//Generate new random code
 	const code = crypto.randomBytes(20).toString('hex');
 
-	//Code is expire in 24 hours
-	return client.setexAsync(`ResetPasswordCode:${code}`, 60 * 60 * 24, userId)
+	//Code is expire in 5 mins
+	return client.setexAsync(`ResetPasswordCode:${code}`, 60 * 5, userId)
 		.then(() => {
 			return code;
```
Comment (abeikverdi) [inline tools/security.js:158]: you dont need to add a catch here. this promise should be caught in a higher level call. Follow the rest of the code. A simple return is sufficient here. Also double check redis to make sure it actually expires @hoan-bitholla

**Review by brandonkimmmm — CHANGES_REQUESTED** [review]

File: tools/security.js [hunk 4]
```diff
@@ -138,32 +142,29 @@ const changeUserPassword = (email, oldPassword, newPassword) => {
...
+
+	//Code is expire in 24 hours
+	return client.setexAsync(`ResetPasswordCode:${code}`, 60 * 60 * 24, userId)
+		.then(() => {
+			return code;
 		})
-		.then((code) => code.code);
+		.catch(() => {
+			throw new Error();
+		});
```
[the remark this answers is not in the document]
Comment (hoan-bitholla) [reply tools/security.js]: I removed it in previous commit

File: tools/security.js [hunk 5]
```diff
@@ -110,7 +111,32 @@ const resetUserPassword = (resetPasswordCode, newPassword) => {
...
 
-const changeUserPassword = (email, oldPassword, newPassword) => {
+const confirmChangeUserPassword = (code, domain) => {
+	return getChangePasswordCode(code)
+		.then((data) => {
+			const dataValues = JSON.parse(data);
+			return all([
+				dbQuery.findOne('user', { where: { id: dataValues.id } }),
+				dataValues,
+				// client.delAsync(`ChangePasswordCode:${code}`)
```
[the remark this answers is not in the document]
Comment (hoan-bitholla) [reply tools/security.js]: This code is outdated! I fixed it!

File: tools/security.js [hunk 6]
```diff
@@ -122,17 +148,51 @@ const changeUserPassword = (email, oldPassword, newPassword) => {
...
+
+const getChangePasswordCode = (code) => {
+	return client.getAsync(`ChangePasswordCode:${code}`)
+		.then((user_id) => {
+			if (!user_id) {
+				const error = new Error(CODE_NOT_FOUND);
+				error.status = 404;
+				throw error;
 			}
+			return user_id;
```
[the remark this answers is not in the document]
Comment (hoan-bitholla) [reply tools/security.js]: I renamed

File: tools/security.js [hunk 7]
```diff
@@ -122,17 +148,51 @@ const changeUserPassword = (email, oldPassword, newPassword) => {
...
+			return client.setexAsync(`ChangePasswordCode:${code}`, 60 * 5, JSON.stringify({
+				id: userId,
+				password: hashedPassword
+			}));
 		})
-		.then((user) => {
-			return user.update({ password: newPassword });
+		.then(() => {
+			return client.ttlAsync(`ChangePasswordCode:${code}`);
+		})
```
[the remark this answers is not in the document]
Comment (hoan-bitholla) [reply tools/security.js]: Because Ali want to double check redis to make sure it acctually expires, so I get TTL of key

**Review by brandonkimmmm — APPROVED** [review]

**Pull request merged**